agents/subtask_predictor.py

```diff
@@ -0,0 +1,311 @@
+            subtask_loss = th.mean(subtask_loss*loss_mask)
+            # calculate total loss
+            tot_loss += th.mean(subtask_loss)
+            metrics[f'p{i}_subtask_loss'] = subtask_loss.item()
+            # get best action from softmax output
+            pred_subtask_indices = th.softmax(pred_subtask, dim=1)
+            pred_subtask_indices = th.argmax(pred_subtask_indices, dim=1)
+            metrics[f'p{i}_subtask_acc'] = accuracy_score(next_subtask[:,i], pred_subtask_indices)
+            tot_loss.backward()
+            self.optimizers[i].step()
+        return metrics
+
+    def evaluate_on_batch(self, batch):
+        """
+        Evaluate agent on 
+        :param batch: The dataset we want to test on
+        :return: fraction of correctly classified samples out of the whole batch
+        """
+        batch = {k: v.to(self.device) for k,v in batch.items()}
+        vo, ao, action, subtasks = batch['visual_obs'].float(), batch['agent_obs'].float(), \
+                                   batch['joint_action'].long(), batch['subtasks'].long()
+
+        curr_subtask, next_subtask = subtasks[:, 0], subtasks[:, 1]
+        metrics = {}
+        for i in range(self.num_players):
+            cs_i = F.one_hot(curr_subtask[:,i], num_classes=Subtasks.NUM_SUBTASKS)
+            sub_history = []
+            # create subtask history of past 10 subtasks
+            for idx, sample in enumerate(cs_i):
+                past_history = []
+                if idx>10:
+                    # append past 10 subtasks 
+                    for ids in range(idx-10,idx):
+                        past_history.append(cs_i[ids])
+                else:
+                    for ids in range(0,10):
+                        past_history.append( [0]*12)
+                #flatten
+                past_history = [item for sublist in past_history for item in sublist]
+                
+                past_history = th.FloatTensor(past_history)
+                sub_history.append(past_history)
+            his = th.empty((batch['agent_obs'].size()[0],120))
+            for idx,sample in enumerate(sub_history):
+                his[idx]= sample
+            pred_subtask = self.players[i].forward( (vo[:,i], ao[:,i], cs_i), his, self.hidden[i] )
+            #flatten
+            self.hidden[i] = pred_subtask
+            # get best action from softmax output
+            pred_subtask_indices = th.softmax(pred_subtask, dim=1)
+            pred_subtask_indices = th.argmax(pred_subtask_indices, dim=1)
+            metrics[f'p{i}_subtask_acc_evaluation'] = accuracy_score(next_subtask[:,i], pred_subtask_indices)
+        return metrics
+
+    def evaluate(self, num_trials):
+        rand_sampler = torch.utils.data.RandomSampler(self.dataset, num_samples=num_trials, replacement=True)
+        dataloader = torch.utils.data.DataLoader(self.dataset, batch_size=num_trials, sampler=rand_sampler)
```

Comment:
You are still evaluating on the training set. To avoid confusion, you should just calculate the metrics from the output of the training and simply log the training metrics (which is essentially what this is)